### PR TITLE
move apiserver Configuration to k8s.io/apiserver/pkg/apis/config

### PR DESCRIPTION
**Pull request by hanxiaoshuai**
**What this PR does / why we need it**:
ref [#2354](https://github.com/kubernetes/community/pull/2354)
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
After the related componentconfig  to be moved to staging, LeaderElectionConfiguration and DebuggingConfiguration should be clean up in pkg/apis/componentconfig
**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by luxas — COMMENTED** [review]
Thanks for this PR!
@rosti FYI
Please add OWNERS file as in the KEP. Fix verify. Add comment to DebuggingConfiguration. Lowercase the EnableContentionProfiling comment. Maybe add validation now or in a later PR. Move/duplicate the "endpoints" string to this package.

**Comment by luxas** [comment]
Feel free to add these packages that need it to .golint_failures btw

**Comment by neolit123** [comment]
@kubernetes/sig-api-machinery-pr-reviews

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/types.go
Comment (stewart-yu) [inline staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/types.go]: IMO, we should add `probuf` for all `var` , syncing with others `api struct`, so that prase easy . such as https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/apis/example/v1/types.go#L60. @luxas  @sttts  WDYT

**Comment by hanxiaoshuai** [comment]
@luxas Done, and I will add validation later if it is need.
/assign @luxas @sttts

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/types.go
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/types.go]: do we support protobuf for configs? Do we want to?

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/doc.go
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/doc.go]: do we want this in openapi?

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/register.go
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/register.go]: remove TODO

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/types.go
Comment (stewart-yu) [reply staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/types.go]: Actually, no need for now,  `Json` enough.

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/doc.go
Comment (hanxiaoshuai) [reply staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/doc.go]: It seems we don't want it, I remove it

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/register.go
Comment (hanxiaoshuai) [reply staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/register.go]: Done, Thanks for your review

File: staging/src/k8s.io/apiserver/pkg/apis/config/OWNERS
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/config/OWNERS]: the OWNERS file from `k8s.io/apiserver` would probably be a more appropriate starting point

File: staging/src/k8s.io/apiserver/pkg/apis/config/types.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: does this belong in apiserver? lots of things that aren't apiservers use leader election. I expected it in apimachinery

Comment (hanxiaoshuai) [reply staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: As it discussed it in  [doc](https://docs.google.com/document/d/1-u2y03ufX7FzBDWv9dVI_HyiIQZL8iz3o3vabeOpP5Y/edit#), in apiserver part, LeaderElectionConfiguration was in it. 
@luxas @sttts can you show some info?

File: staging/src/k8s.io/apiserver/pkg/apis/config/OWNERS
Comment (hanxiaoshuai) [reply staging/src/k8s.io/apiserver/pkg/apis/config/OWNERS]: @luxas @sttts what's your opinion?

**Review by luxas — COMMENTED** [review]
Commenting here as I can comment inline:
- No need for protobuf at least now. Let's get the normal config part done first and then start thinking about some apiserver supportin serving the API via protobuf.
- Do we have an alias for k8s.io/apiserver reviewers? If not that should be created and referenced as @liggitt pointed out. Also, the api-approvers/reviewers OWNERS teams should be referenced.
@sttts pointed out aggregated API servers might need it leader election stuff in the future? IDK, if not we can move that to k8s.io/controller-manager or k8s.io/apimachinery if you think that makes more sense

**Comment by luxas** [comment]
...as I can't comment inline from mobile ;)

**Comment by luxas** [comment]
/assign @thockin @liggitt
Waiting for their approval as well, and the KEP to merge before this PR should merge

File: staging/src/k8s.io/apiserver/pkg/apis/config/OWNERS
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/apis/config/OWNERS]: +1 for @liggitt's suggestion.

File: staging/src/k8s.io/apiserver/pkg/apis/config/types.go
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: what are the value of this? resource plus apigroup plus version?

File: staging/src/k8s.io/apiserver/pkg/apis/config/types.go
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: We were undecided about this one. It seems to be higher level than apimachinery, but also has no tight connections to apiserver. I don't feel strongly with either direction. As the types are kube independent, we can technically also move it to apimachinery. I would like to see some generic `ResourceLock` syntax which does not reference kube.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: cc @mtaufen @wojtek-t who were working on the Lease API

Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: > does this belong in apiserver? lots of things that aren't apiservers use leader election. I expected it in apimachinery

I don't see this as related to our type system in the same way as `TypeMeta`, `ObjectMeta`, `Time`, or `List`.  I don't even see it as quite as fundamental as `ListOptions`.  Something common for many bits of config, sure.  Something common for all types/schemes, not really.

Comment (mtaufen) [reply staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: If I understand correctly, components can migrate to using the Lease API (it was designed to represent the same information as LeaderElectionRecord), but they still have to get related configuration from somewhere (i.e. LeaderElectionConfiguration).

So this is a question of where to put a common config struct type?

File: staging/src/k8s.io/apiserver/pkg/apis/config/types.go
Comment (roycaihw) [reply staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: The default is string "endpoints": https://github.com/kubernetes/kubernetes/pull/66059/files#diff-315017a8fe6dd5515919e19a0fffcc20R43

Comment (roycaihw) [reply staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: There are two lock types `endpoints` and `configmaps`: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/interface.go

**Comment by sttts** [comment]
Can we at least create the clean-up PR now already to see possible issues early? Don't like to duplicate code and post-poin the cleanups.

/hold

**Comment by luxas** [comment]
@sttts The PR that cleans this up/moves the usage is here: #66722.
I prefer having clean PRs like this so that one PR adds the net-new code and one PR cleans stuff up later if that is ok by you.
/hold cancel

File: staging/src/k8s.io/apiserver/pkg/apis/config/types.go
Comment (thockin) [reply staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: is this resolved?

**Comment by luxas** [comment]
Can we please get a top-level approval of this to get the ball rolling so we can then cleanup the types (PR https://github.com/kubernetes/kubernetes/pull/66722) ASAP, and unblock the components' types moving (e.g. https://github.com/kubernetes/kubernetes/pull/66916).

We can always change the structure later if we feel we need to, before going beta with this, but if this is gonna be stuck forever we have no chance improving anything.

**Comment by jbeda** [comment]
/approve

Just to get this moving.

File: staging/src/k8s.io/apiserver/pkg/apis/config/types.go
Comment (luxas) [reply staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: Yes, I think the consensus is to put it here, in k8s.io/apiserver vs putting it in k8s.io/apimachinery which should hold even more generic types.

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/doc.go
Comment (luxas) [inline staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/doc.go]: As said in the other PR, we're not sure if we actually need this, but we're leaving it as-is for now

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/defaults.go
Comment (luxas) [inline staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/defaults.go]: nit: in the future you can use `k8s.io/utils/pointer.BoolPtr(true)`

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/defaults.go
Comment (luxas) [inline staging/src/k8s.io/apiserver/pkg/apis/config/v1alpha1/defaults.go]: TODO: we haven't fully decided if we're gonna register the defaults like this in the scheme, but having this as it is the default way of doing things right now. We can always change stuff later.

**Review by luxas — APPROVED** [review]
/lgtm

Thanks @hanxiaoshuai and @jbeda!
Let's get the ball rolling here and follow-up on minor nits and changes as we go.

**Pull request merged**
